clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java

```diff
@@ -0,0 +1,303 @@
+                              final long requestTimeoutMs,
+                              final CoordinatorRequestState coordinatorRequestState) {
+        Objects.requireNonNull(groupId);
+        this.time = time;
+        this.log = logContext.logger(this.getClass());
+        this.errorHandler = errorHandler;
+        this.groupId = groupId;
+        this.rebalanceTimeoutMs = rebalanceTimeoutMs;
+        this.requestTimeoutMs = requestTimeoutMs;
+        this.coordinatorRequestState = coordinatorRequestState;
+    }
+
+    @Override
+    public NetworkClientDelegate.PollResult poll(final long currentTimeMs) {
+        if (this.coordinator != null) {
+            return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, Collections.emptyList());
+        }
+
+        if (coordinatorRequestState.canSendRequest(currentTimeMs)) {
+            NetworkClientDelegate.UnsentRequest request = makeFindCoordinatorRequest(currentTimeMs);
+            return new NetworkClientDelegate.PollResult(0, Collections.singletonList(request));
+        }
+
+        return new NetworkClientDelegate.PollResult(
+                coordinatorRequestState.remainingBackoffMs(currentTimeMs),
+                new ArrayList<>());
+    }
+
+    private NetworkClientDelegate.UnsentRequest makeFindCoordinatorRequest(final long currentTimeMs) {
+        coordinatorRequestState.updateLastSend(currentTimeMs);
+        FindCoordinatorRequestData data = new FindCoordinatorRequestData()
+                .setKeyType(FindCoordinatorRequest.CoordinatorType.GROUP.id())
+                .setKey(this.groupId);
+        return NetworkClientDelegate.UnsentRequest.makeUnsentRequest(
+                this.time.timer(requestTimeoutMs),
+                new FindCoordinatorRequest.Builder(data),
+                new FindCoordinatorRequestHandler());
+    }
+
+    /**
+     * Mark the current coordinator null and return the old coordinator. Return an empty Optional
+     * if the current coordinator is unknown.
+     * @param cause why the coordinator is marked unknown
+     * @return Optional coordinator node that can be null.
+     */
+    protected void markCoordinatorUnknown(final String cause, final long currentTimeMs) {
+        if (this.coordinator != null) {
+            log.info("Group coordinator {} is unavailable or invalid due to cause: {}. "
+                            + "Rediscovery will be attempted.", this.coordinator, cause);
+            this.coordinator = null;
+            timeMarkedUnknownMs = currentTimeMs;
+        } else {
+            long durationOfOngoingDisconnect = Math.max(0, currentTimeMs - timeMarkedUnknownMs);
+            if (durationOfOngoingDisconnect > this.rebalanceTimeoutMs)
+                log.debug("Consumer has been disconnected from the group coordinator for {}ms",
+                        durationOfOngoingDisconnect);
+        }
+    }
+
+    private void handleSuccessFindCoordinatorResponse(
```

Comment:
I think we can make these names more concise. The request type is already clear in the context, so we could call this `onSuccessfulResponse` for example.